### DISCUSSION TITLE
data persistence vs state input

### DISCUSSION THREAD

**Discussion by dennissetiawan** [indeedeng/iwf#277] [Q&A]
When should we use data persistence vs passing the data through state input?

**Comment by longquanzheng** [comment]
Either will work. 
Some tips for the decisions:
* if there will be only one state needs the value, then it’s easier to use state
* if it’s more than one, you probably want to use data or search attributes to save the code of passing values around .
* By default , all state api / RPC will load all data attributes unless you set a different persistence loading policy. So Having too many data attributes could be wasteful for communication for default loading, and could exceed the 2MB limit . So if you can use input and it’s easier , then use input . 
* You can set search attributes in start workflow api without a starting state(starting state is not required, not event a state ) . However, data attributes is not supported in start api yet. We may support data attributes as well in the future (it’s a small work but we haven’t found use case to prioritize it )
